delocate/tools.py

```
@@ -272,14 +407,48 @@ def get_install_id(filename):
     install_id : str
         install id of library `filename`, or None if no install id
     """
-    lines = _cmd_out_err(["otool", "-D", filename])
-    if not _line0_says_object(lines[0], filename):
-        return None
-    if len(lines) == 1:
-        return None
-    if len(lines) != 2:
-        raise InstallNameError("Unexpected otool output " + "\n".join(lines))
-    return lines[1].strip()
+    install_ids = _get_install_ids(filename)
+    if all(not install_id for install_id in install_ids.values()):
+        return None  # No install ids or nothing returned.
+    my_id = list(install_ids.values())[0]
+    if any(install_id != my_id for install_id in install_ids.values()):
+        raise InstallNameError(
+            "This function does not support separate install ids"
+            f" per-architecture: {install_ids}"
+        )
+    return my_id  # Only one install name.
+
+
+def _get_install_ids(filename: str) -> Dict[str, str]:
+    """Return the install ids of a library.
+
+    Parameters
+    ----------
+    filename : str
+        filename of library
+
+    Returns
+    -------
+    install_ids : dict
+        install ids of library `filename`.
+        The key is the architecture which is '' if none is provided by otool.
+        The value is the install id for that architecture.
+        If the library has no install ids then an empty dict is returned.
+    """
+    otool = subprocess.run(
+        ["otool", "-D", filename],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        universal_newlines=True,
+    )
+    if not _line0_says_object(otool.stdout or otool.stderr, filename):
+        return {}
+    out = {}
+    for arch, my_id_list in _parse_otool_listing(otool.stdout).items():
+        if not my_id_list:
+            continue
+        (out[arch],) = my_id_list  # List should always have 0 or 1 items.
```

How about:

```suggestion
        if my_id_list:
            out[arch] = my_id_list[0]  # List should always have 0 or 1 items.
```

---

This code `(out[arch],) = my_id_list` is actually checking that the list only has one item, raising a TypeError if it has any other number of items.

`out[arch] = my_id_list[0]` would silently discard extra items, only checking that the list isn't empty.

It's possible that this is an unnecessary sanity check.  I'm not sure what to expect in the future.

---

OK - but then I think, make it explicit that you are doing that check, rather than relying on the reader to infer it.  Maybe something like:

How about:

```suggestion
        if my_id_list:
            if len(my_id_list) > 1:
                raise InstallNameError("List should always have 0 or 1 items.")
            out[arch] = my_id_list[0]
```